File: sdk/dotnet/src/Snaploom.Capture/CaptureTypes.cs
namespace Snaploom.Capture;

/// <summary>Options used when creating a capture client.</summary>
public sealed class CaptureClientOptions
{
    /// <summary>An optional absolute path to a separately installed Capture Host.</summary>
    public string? HostExecutableOverride { get; init; }

    /// <summary>An optional Host launch timeout. Native defaults apply when omitted.</summary>
    public TimeSpan? LaunchTimeout { get; init; }

    /// <summary>An optional Host handshake timeout. Native defaults apply when omitted.</summary>
    public TimeSpan? HandshakeTimeout { get; init; }
}

/// <summary>Options for one interactive capture.</summary>
public sealed class CaptureOptions
{
    /// <summary>Disables the default clipboard write when true.</summary>
    public bool DisableClipboard { get; init; }

    /// <summary>An optional interaction timeout. No deadline is imposed when omitted.</summary>
    public TimeSpan? InteractionTimeout { get; init; }
}

/// <summary>The loaded native SDK version.</summary>
/// <param name="AbiMajor">The C ABI major version.</param>
/// <param name="NativeSemver">The native SDK semantic version.</param>
public sealed record CaptureSdkVersion(uint AbiMajor, string NativeSemver);

/// <summary>An owned capture result. PNG bytes no longer reference native memory.</summary>
/// <param name="Png">The encoded PNG bytes.</param>
/// <param name="PixelWidth">The physical pixel width.</param>
/// <param name="PixelHeight">The physical pixel height.</param>
/// <param name="ClipboardWritten">Whether the same PNG bytes were written to the clipboard.</param>
public sealed record CaptureResult(
    ReadOnlyMemory<byte> Png,
    uint PixelWidth,
    uint PixelHeight,
    bool ClipboardWritten);

/// <summary>Stable errors reported asynchronously by the Capture SDK.</summary>
public enum CaptureErrorCode : uint
{
    /// <summary>No error.</summary>
    None = 0,
    /// <summary>Another global capture session is active.</summary>
    Busy = 1,
    /// <summary>The separately installed Capture Host was not found.</summary>
    HostNotFound = 2,
    /// <summary>The Capture Host could not be started.</summary>
    HostStartFailed = 3,
    /// <summary>Starting the Capture Host timed out.</summary>
    HostStartTimeout = 4,
    /// <summary>Peer authentication failed.</summary>
    AuthenticationFailed = 5,
    /// <summary>The Host and SDK protocol versions are incompatible.</summary>
    ProtocolIncompatible = 6,
    /// <summary>The Host sent an invalid protocol message.</summary>
    ProtocolError = 7,
    /// <summary>The Capture Host exited during the request.</summary>
    HostCrashed = 8,
    /// <summary>The local IPC transport failed.</summary>
    TransportFailed = 9,
    /// <summary>The authenticated handshake timed out.</summary>
    HandshakeTimeout = 10,
    /// <summary>The overall request timed out.</summary>
    RequestTimeout = 11,
    /// <summary>Capture is unavailable on the current platform.</summary>
    PlatformUnavailable = 20,
    /// <summary>The required screen recording permission was not granted.</summary>
    PermissionNotGranted = 21,
    /// <summary>A previously granted permission was revoked.</summary>
    PermissionRevoked = 22,
    /// <summary>The requested display is no longer available.</summary>
    DisplayUnavailable = 23,
    /// <summary>The platform capture service is unavailable.</summary>
    CaptureUnavailable = 24,
    /// <summary>Platform frame capture timed out.</summary>
    CaptureTimeout = 25,
    /// <summary>Captured pixels could not be converted.</summary>
    PixelConversionFailed = 26,
    /// <summary>The returned image violated result invariants.</summary>
    InvalidResult = 30,
    /// <summary>The returned image exceeded the 128 MiB contract limit.</summary>
    ResultTooLarge = 31,
    /// <summary>Writing the final PNG to the clipboard failed.</summary>
    ClipboardWriteFailed = 32,
    /// <summary>The native SDK could not allocate memory.</summary>
    OutOfMemory = 40,
    /// <summary>The client was already closed.</summary>
    ClientClosed = 41,
    /// <summary>An internal native error occurred.</summary>
    Internal = 255,
}

/// <summary>Base exception for Capture SDK protocol and native status failures.</summary>
public class CaptureSdkException : Exception
{
    internal CaptureSdkException(string message, uint? nativeStatus = null, Exception? innerException = null)
        : base(message, innerException)
    {
        NativeStatus = nativeStatus;
    }

    /// <summary>The raw synchronous native status, when the failure came from a C call.</summary>
    public uint? NativeStatus { get; }
}

/// <summary>An asynchronous capture failure with a stable raw error value.</summary>
public sealed class CaptureException : CaptureSdkException
{
    internal CaptureException(CaptureErrorCode errorCode, bool isRetryable, string diagnosticName)
        : base($"Snaploom capture failed: {diagnosticName} ({(uint)errorCode}).")
    {
        ErrorCode = errorCode;
        IsRetryable = isRetryable;
    }

    /// <summary>The stable error value. Unknown future values remain representable.</summary>
    public CaptureErrorCode ErrorCode { get; }

    /// <summary>Whether a later, newly initiated request may succeed.</summary>
    public bool IsRetryable { get; }
}

/// <summary>The user canceled from the Capture Host UI rather than through a caller token.</summary>
public sealed class CaptureCanceledException : CaptureSdkException
{
    internal CaptureCanceledException()
        : base("The capture was canceled in the Snaploom Capture Host.")
    {
    }
}

/// <summary>A synchronous argument or ABI layout status reported by native code.</summary>
public sealed class CaptureArgumentException : ArgumentException
{
    internal CaptureArgumentException(uint nativeStatus)
        : base($"The Snaploom Capture SDK rejected an argument (native status {nativeStatus}).")
    {
        NativeStatus = nativeStatus;
    }

    /// <summary>The raw synchronous native status.</summary>
    public uint NativeStatus { get; }
}

File: sdk/swift/Sources/SnaploomCapture/CaptureTypes.swift
import Foundation

/// Options used when creating a capture client.
public struct CaptureClientOptions: Sendable, Equatable {
  /// Optional absolute path to a separately installed Capture Host.
  public var hostExecutableOverride: String?

  /// Optional Host launch timeout in milliseconds. Native defaults apply when omitted.
  public var launchTimeoutMilliseconds: UInt32?

  /// Optional Host handshake timeout in milliseconds. Native defaults apply when omitted.
  public var handshakeTimeoutMilliseconds: UInt32?

  public init(
    hostExecutableOverride: String? = nil,
    launchTimeoutMilliseconds: UInt32? = nil,
    handshakeTimeoutMilliseconds: UInt32? = nil
  ) {
    self.hostExecutableOverride = hostExecutableOverride
    self.launchTimeoutMilliseconds = launchTimeoutMilliseconds
    self.handshakeTimeoutMilliseconds = handshakeTimeoutMilliseconds
  }
}

/// Options for one interactive capture.
public struct CaptureOptions: Sendable, Equatable {
  /// Disables the default clipboard write when true.
  public var disableClipboard: Bool

  /// Optional interaction timeout in milliseconds. Native defaults apply when omitted.
  public var interactionTimeoutMilliseconds: UInt64?

  public init(
    disableClipboard: Bool = false,
    interactionTimeoutMilliseconds: UInt64? = nil
  ) {
    self.disableClipboard = disableClipboard
    self.interactionTimeoutMilliseconds = interactionTimeoutMilliseconds
  }
}

/// The ABI and semantic version reported by the loaded native SDK.
public struct CaptureSDKVersion: Sendable, Equatable {
  public let abiMajor: UInt32
  public let nativeSemver: String

  public init(abiMajor: UInt32, nativeSemver: String) {
    self.abiMajor = abiMajor
    self.nativeSemver = nativeSemver
  }
}

/// An owned capture result. PNG bytes no longer reference native memory.
public struct CaptureResult: Sendable, Equatable {
  public let pngData: Data
  public let pixelWidth: UInt32
  public let pixelHeight: UInt32
  public let clipboardWritten: Bool

  public init(
    pngData: Data,
    pixelWidth: UInt32,
    pixelHeight: UInt32,
    clipboardWritten: Bool
  ) {
    self.pngData = pngData
    self.pixelWidth = pixelWidth
    self.pixelHeight = pixelHeight
    self.clipboardWritten = clipboardWritten
  }
}

/// A forward-compatible stable native error value.
public struct CaptureErrorCode: RawRepresentable, Sendable, Hashable {
  public let rawValue: UInt32

  public init(rawValue: UInt32) {
    self.rawValue = rawValue
  }

  public static let none = Self(rawValue: 0)
  public static let busy = Self(rawValue: 1)
  public static let hostNotFound = Self(rawValue: 2)
  public static let hostStartFailed = Self(rawValue: 3)
  public static let hostStartTimeout = Self(rawValue: 4)
  public static let authenticationFailed = Self(rawValue: 5)
  public static let protocolIncompatible = Self(rawValue: 6)
  public static let protocolError = Self(rawValue: 7)
  public static let hostCrashed = Self(rawValue: 8)
  public static let transportFailed = Self(rawValue: 9)
  public static let handshakeTimeout = Self(rawValue: 10)
  public static let requestTimeout = Self(rawValue: 11)
  public static let platformUnavailable = Self(rawValue: 20)
  public static let permissionNotGranted = Self(rawValue: 21)
  public static let permissionRevoked = Self(rawValue: 22)
  public static let displayUnavailable = Self(rawValue: 23)
  public static let captureUnavailable = Self(rawValue: 24)
  public static let captureTimeout = Self(rawValue: 25)
  public static let pixelConversionFailed = Self(rawValue: 26)
  public static let invalidResult = Self(rawValue: 30)
  public static let resultTooLarge = Self(rawValue: 31)
  public static let clipboardWriteFailed = Self(rawValue: 32)
  public static let outOfMemory = Self(rawValue: 40)
  public static let clientClosed = Self(rawValue: 41)
  public static let `internal` = Self(rawValue: 255)

  /// A stable diagnostic name. Unknown future values remain representable.
  public var diagnosticName: String {
    switch self {
    case .none: "NONE"
    case .busy: "BUSY"
    case .hostNotFound: "HOST_NOT_FOUND"
    case .hostStartFailed: "HOST_START_FAILED"
    case .hostStartTimeout: "HOST_START_TIMEOUT"
    case .authenticationFailed: "AUTHENTICATION_FAILED"
    case .protocolIncompatible: "PROTOCOL_INCOMPATIBLE"
    case .protocolError: "PROTOCOL_ERROR"
    case .hostCrashed: "HOST_CRASHED"
    case .transportFailed: "TRANSPORT_FAILED"
    case .handshakeTimeout: "HANDSHAKE_TIMEOUT"
    case .requestTimeout: "REQUEST_TIMEOUT"
    case .platformUnavailable: "PLATFORM_UNAVAILABLE"
    case .permissionNotGranted: "PERMISSION_NOT_GRANTED"
    case .permissionRevoked: "PERMISSION_REVOKED"
    case .displayUnavailable: "DISPLAY_UNAVAILABLE"
    case .captureUnavailable: "CAPTURE_UNAVAILABLE"
    case .captureTimeout: "CAPTURE_TIMEOUT"
    case .pixelConversionFailed: "PIXEL_CONVERSION_FAILED"
    case .invalidResult: "INVALID_RESULT"
    case .resultTooLarge: "RESULT_TOO_LARGE"
    case .clipboardWriteFailed: "CLIPBOARD_WRITE_FAILED"
    case .outOfMemory: "OUT_OF_MEMORY"
    case .clientClosed: "CLIENT_CLOSED"
    case .internal: "INTERNAL"
    default: "UNKNOWN"
    }
  }
}

/// An asynchronous native failure with its retryability contract preserved.
public struct CaptureFailure: Error, Sendable, Equatable {
  public let code: CaptureErrorCode
  public let isRetryable: Bool

  public init(code: CaptureErrorCode, isRetryable: Bool) {
    self.code = code
    self.isRetryable = isRetryable
  }
}

/// Synchronous SDK, ABI, cancellation, and result-validation failures.
public enum CaptureError: Error, Sendable, Equatable {
  case invalidArgument(status: UInt32)
  case clientClosed
  case outOfMemory
  case nativeStatus(rawValue: UInt32)
  case incompatibleNative(reason: String)
  case canceledByUser
  case failure(CaptureFailure)
  case invalidCompletion(reason: String)
  case unknownCompletion(rawValue: UInt32)
}
